File: Linear/Static/Array/06_Sorting/Selection Sort/SelectionSort.c
#include <stdio.h>
#include <unistd.h>

// An optimized version of Selection Sort
void selectionSort(int arr[], int n) {
    for (int i = 0; i < n - 1; i++) {

        /* Assume the current position
        holds the minimum element */
        int indexMin = i;

        /* Iterate through the unsorted
        portion to find the actual minimum */
        for (int j = i; j < n; j++) {
            if (arr[j] < arr[indexMin])
            indexMin = j;   // Update indexMin if a smaller element is found
        }

        /* Move minimum element
        to its correct position */
        int temp = arr[i];
        arr[i] = arr[indexMin];
        arr[indexMin] = temp;
    }
}

// Function to print an array
void printArray(int arr[], int n) {
    for (int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }    
}

int main() {
    int arr[] = { 64, 25, 12, 22, 11 };
    int n = sizeof(arr) / sizeof(arr[0]);
    printf("Unsorted array: \n");
    printArray(arr, n);

    for (int i = 3; i > 0; i--) {
        printf("\nSorting in %d", i);
        sleep(1);
    }

    selectionSort(arr, n);
    printf("\nSorted array: \n");
    printArray(arr, n);

    return 0;
}

File: Linear/Static/Array/06_Sorting/Selection Sort/SelectionSort.py
import time

# An optimized version of Selection Sort
def selectionSort(arr):
    n = len(arr)

    for i in range(n - 1):

        """ Assume the current position
        holds the minimum element """
        indexMin = i

        """ Iterate through the unsorted
        portion to find the actual minimum """
        for j in range(i, n):
            if arr[j] < arr[indexMin]:
                indexMin = j    # Update indexMin if a smaller element is found
        """ Iterate through the unsorted
        portion to find the actual minimum """
        arr[i], arr[indexMin] = arr[indexMin], arr[i]
    return arr

# Function to print an array
def printArray(arr):
    for i in range(len(arr)):
        print(arr[i], end=" ")

arr = [ 64, 25, 12, 22, 11 ]
print("Unsorted array:")
printArray(arr)

for i in range(3, 0, -1):
    print("\nSorting in", i)
    time.sleep(1)

arr = selectionSort(arr)
print("Sorted array:")
printArray(arr)